### PR TITLE
Don't expose HTTP API for secure clusters

### LINKED ISSUE #6407: Don't expose HTTP API for secure clusters
https://github.com/dask/distributed/pull/6270 exposed a new HTTP API, enabled by default. Copied from https://github.com/dask/distributed/pull/6270#issuecomment-1133358073:

I'm concerned about a security regression here. By default, this is opening up an API that allows anyone to change cluster state (via `retire_workers` currently, but I imagine other things might be added someday too).

Prior to this, the only way to do things that affected cluster state was through the client. All the HTTP routes were effectively read-only. (Whether there is a vulnerability in the bokeh dashboard is another topic; it's pretty possible there is, but I'm just talking here in principle.)

I think it's rather common to expose the HTTP routes to the public internet. For example, I believe dask-cloudprovider does this:
> By default a Dask security group will be created with ports 8786 and 8787 exposed to the internet https://cloudprovider.dask.org/en/latest/aws.html#dask_cloudprovider.aws.EC2Cluster

You want those ports exposed for convenience, so you can connect to them. But you don't want anyone to be able to do stuff to the cluster, so you set up TLS using temporary credentials. dask-cloudprovider does this for you as well:
> When a cluster is launched with any of these cluster managers a set of temporary keys will be generated and distributed to the cluster nodes via their startup script. All communication between the client, scheduler and workers will then be encrypted and only clients and workers with valid certificates will be able to connect to the scheduler.
https://cloudprovider.dask.org/en/latest/security.html#authentication-and-encryption

Currently, if you set up TLS for your cluster, this is [mTLS](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/), meaning the scheduler verifies the client's certificate ([docs](https://distributed.dask.org/en/stable/tls.html#security-policy), [code](https://github.com/dask/distributed/blob/9bb999d4b66670570d9d8f53d04501dec6a25e7e/distributed/security.py#L275-L276)). This serves as a form of authentication and authorization: if you've set up cluster security, you can only tell the scheduler to do things if you hold a valid certificate.

However, the HTTP routes have no authentication (they use standard TLS, not mTLS, because mTLS would be very inconvenient when you want to look at the dashboard with a web browser).

So after this change, someone who had gone to the trouble to set up mTLS for their cluster (or was using the defaults of their cluster deployment system) would, by default, have an unauthenticated endpoint running that allowed anyone with access to `:8787` (aka the dashboard) to affect cluster state.

----

I think we should do two things:
1. Short-term: disable the HTTP API if TLS is specified on the scheduler. This is a reasonable default.
2. Long-term: figure out a security posture and authentication for the HTTP API that's consistent with the security posture of other things that can affect cluster state (aka the client).

### PR DESCRIPTION
**Pull request by gjoseph92**
Closes #6407

I'd like to come up with a better name than `insecure-routes` for the config, since that implies that the others are secure (I don't know if that's something we really guarantee?)

cc @Matt711 @jakirkham 

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @jacobtomlinson @quasiben (for awareness)

File: distributed/http/scheduler/tests/test_scheduler_http.py [hunk 1]
```diff
@@ -259,6 +259,15 @@ async def test_api(c, s, a, b):
             assert (await resp.text()) == "API V1"
 
 
+@gen_cluster(client=True, clean_kwargs={"threads": False}, security=tls_only_security())
+async def test_api_disabled_if_secure(c, s, a, b):
+    async with aiohttp.ClientSession() as session:
+        async with session.get(
+            "http://localhost:%d/api/v1" % s.http_server.port
+        ) as resp:
+            assert resp.status == 404
```
Comment (fjetter) [inline distributed/http/scheduler/tests/test_scheduler_http.py:268]: if there is a typo, the route is actually removed or the version number is bumped, this test would still pass.
You could parametrize over security and ensure that the route is reachable if security is disabled to make sure the test assumption is actually valid

File: docs/source/http_services.rst [hunk 2]
```diff
@@ -52,7 +52,11 @@ Pages and JSON endpoints served by the scheduler
 Scheduler API
 -------------
 
-Scheduler methods exposed by the API with an example of the request body they take
+Scheduler methods exposed by the API with an example of the request body they take.
+
+.. note::
+   To prevent unauthorized access, the scheduler API is disabled by default if `tls`_ is enabled.
+   See the ``distributed.http.insecure-routes`` :doc:`config <configuration>` setting.
```
Comment (fjetter) [inline docs/source/http_services.rst:59]: The API is disabled if `tls` is **disabled**. This sentence is the other way round, isn't it?

**Review by fjetter — COMMENTED** [review]
> I'd like to come up with a better name than insecure-routes for the config,

How about `tls-required-routes`?

**Comment by fjetter** [comment]
slight off-topic: Is HTTPS something that would help? Is this something we can easily do? Are there reasons why we wouldn't want to do HTTPS? (We should create a follow up ticket if the answer is yes, we want to do this)

**Review by jacobtomlinson — COMMENTED** [review]
Thanks for raising this @gjoseph92. I agree that we should disable the HTTP API when mTLS is enabled. I would also be happy to just disable the HTTP API by default and make it opt-in.

I'm not especially excited about splitting HTTP routes into "secure" and "insecure" like this. It feels like we are leaking implementation details to the user here. We should definitely add some auth options to the API in the future, so then we would be left with a stray "insecure" config option that no longer makes sense.

**Comment by gjoseph92** [comment]
> We should definitely add some auth options to the API in the future, so then we would be left with a stray "insecure" config option that no longer makes sense

Yeah, I was hesitant about that too.

@jacobtomlinson @Matt711 would you prefer just leaving `distributed.http.scheduler.api` out of the default routes for now then?

**Comment by gjoseph92** [comment]
Sounds good. @jacobtomlinson would someone on your side be able to pick that up today, or should I? (I'd want to see a test confirming that the API is not enabled with the default config to avoid regressions FYI.) We're hoping to get the release out today. I'll close out this PR then.

**Comment by jakirkham** [comment]
My guess is Jacob is out for the day (kind of late for him). Matt was an intern, whose last day was Friday. Think we are going to struggle to come up with someone to fix it with short notice (today). If we have some more time, that might be more doable.

**Pull request closed without merging**